### ISSUE TITLE
Authentication support

### ISSUE THREAD

**Issue by andrenth** [netdata/netdata#70]
Hello

Are there plans for supporting some kind of user authentication to the web interface?

Cheers.

**Comment by ktsaou** [comment]
Yes, but it is low on my priorities.
If you have another web server installed, just pass netdata through it and force authentication on the web server. You can find information about passing netdata through another web server, at the wiki.

**Comment by artberri** [comment]
I have added to the wiki how to enable authentication with nginx:

https://github.com/firehol/netdata/wiki/Running-behind-nginx

**Comment by ktsaou** [comment]
> I have added to the wiki how to enable authentication with nginx

Perfect!

**Comment by FingerLiu** [comment]
If you can add a config that tell netdata server **only listen 127.0.0.1** rather than 0.0.0.0,
I think use it under nginx will be OK, and there will be no need to add firewall using iptables.

**Comment by SylvainTI** [comment]
Agreed: security should be by default, an action should be required to expose all these sensitive info to the world.

**Comment by ktsaou** [comment]
A few folks at #49 are already onto it. Waiting for a PR...

**Comment by ktsaou** [comment]
Added an option to bind to an IP address given in netdata config.
Check #49

**Comment by Eadrom** [comment]
I just found netdata today.  I was playing around with it and wanted a way to secure the output from my servers.  I have been setting up port forwarding via SSH to secure my netdata output.  

`ssh -i $SSH_KEY_PATH -fN -L $LOCAL_PORT:localhost:$REMOTE_PORT $USER@$REMOTE_HOST`

Then I just point my browser or custom dashboard at a port(s) local to my laptop or desktop, and all the data is flowing over a key-authenticated SSH connection.  The remote netdata server is running on localhost only port and so isn't visible from the Internet.  I'm working on building a local HTML file that has a nicely put together custom dashboard for my various servers and a service unit for systemd that will respawn failed SSH connections. 

I'd be curious on feedback if there are any security issues with this approach.

**Comment by rileytg** [comment]
I'm curious if @eadrom s solution has any know problems? sounds clever to me

**Comment by jarmo** [comment]
@rileytg I'm using exactly the same solution as @Eadrom. Only problem with this solution is that registry won't work (so you can't select your servers from netdata dashboard), but this is not a big issue for me now.

I've also had some plans to do my own dashboard like @Eadrom, but haven't gotten time to do it yet. And I'm not using anything to respawn failed SSH connections because I only use port forwarding when I want to look at the dashboard and that's not all the time.

Anyway, I don't think that netdata itself should implement all security features - basically only thing it should do (if not doing already) is to bind only on local network interface by default so that it would not accidentally be visible to the internet.

**Comment by ktsaou** [comment]
> Anyway, I don't think that netdata itself should implement all security features - basically only thing it should do (if not doing already) is to bind only on local network interface by default so that it would not accidentally be visible to the internet.

netdata can already bind to one or more ports or all, one or more IPs. Check `netdata.conf`, option `bind to`.

Regarding authentication, what I plan to do is centralize it at the registry, or possibly the forthcoming central health server. So, you are going to say to all your netdata that you want authenticated access and the registry will know how how to do it, centrally for all your netdata servers.

**Comment by jarmo** [comment]
My point about binding to local interface was that it should do it by default and not the other way around. In that way anything isn't exposed accidentally to the Internet and netdata user has to change configuration explicitly to create this hole to their server.

**Comment by ktsaou** [comment]
Let's think of it:

1. The netdata web dashboard is read-only at the moment, so apart from obtaining information about the system, an unauthorized user cannot take any action via netdata. Netdata by default runs as an **unprivileged user** and even in the case of a DDoS, netdata runs with the **IDLE process scheduler** (lower that nice 19) and will be **killed first if the system starves for resources** (it sets its out-of-memory score to the maximum on start).

2. If netdata was binding to `lo` by default, every single user installing netdata at a headless machine (most of us), would have to edit the configuration and change this setting. Frustration.

3. On the other hand, if netdata is accidentally exposed to the internet, the exposed system has much bigger problems: at least its firewall or DMZ policy is completely broken.

So, I find it reasonable to have it bind on all IPs by default.

**Comment by kachkaev** [comment]
I'm running a couple of virtual private servers, which are not a part of any cloud or a corporate network. They are facing the public web and are used to host quite simple web projects. At the moment I see these options to protect netdata from various curious people or bulk scanners of port 19999:

* Change the port to 19998 or any other 'secret' number
* Run netdata behind a shared nginx proxy or apache and manually define HTTP auth there
* Run netdata inside a docker container and place it behind a dockerized nginx. It is even possible to achieve a complete automated security stack including a valid HTTPS cert and HTTP Auth, but this will involve relying on  [jwilder/nginx-proxy](https://github.com/jwilder/nginx-proxy), [JrCs/docker-letsencrypt-nginx-proxy-companion](https://github.com/JrCs/docker-letsencrypt-nginx-proxy-companion) and also to spin-up an extra nginx proxy to configure login and password until https://github.com/jwilder/nginx-proxy/issues/456 is closed.

The first option should to the job, but does not feel to be reliable. The problem with the two latter ones is that they introduce additional entities to care about and each of them can fail. If my shared nginx proxy (either docker or non-docker one) ever crashes, it becomes impossible to see the charts without ssh-ing to the server and fixing the cause (useful numbers in netdata memory can get wiped by then). Besides, the shared reverse-proxy can also be a subject to tests and experiments, which suggests that netdata probably deserves its own reverse proxy. This complicates things even more.

I understand why netdata does not include any auth and also agree that having port 19999 controlled by the firewall is the best solution. But unfortunately, not all VPS provides give this luxury to their clients. In my view, quite a few members of the today's and future netdata audience could benefit from having simple auth built into the app and thus keep the same awesome setup/upgrade experience.

Ultimately, netdata could also optionally auto-generate a self-signed HTTPS certificate for itself to protect the app traffic, which would include HTTP login & password (or maybe some auto token?). I understand that doing so is already possible with a proxy, but can't resist from sharing this thought :–)

Here is my dream in one line:
```bash
./netdata-install.sh --port=12345 --cert=self-signed --http_user=netdata --http_password=rocks
```

:–)

**Comment by ktsaou** [comment]
Nice...

well, I haven't added SSL yet mainly because it will complicate things and I believe I am not ready yet for it. I plan to restructure a lot of netdata internals and I am trying to avoid the complication of SSL on top of it. Also, it will require a lot more build libraries.

Regarding authentication, I agree it is simple to add, but I don't want to add it to each netdata.

I think there should be an OAUTH2 provider, which will be used by all netdata to protect the content, so that accounts will be configured centrally. There should also be something like API tokens for applications or integrations to access netdata with a limited scope (e.g. API token X allows viewing only netdata1, or chart `system.cpu` on all netdata servers, etc). I am not sure how to do this yet (i.e. like the registry - any netdata could be the OAUTH2 provider or a separate daemon that can be maintained independently?)

So, patience...

Keep in mind you currently have a few more options:

1. ssh to the server to access netdata. So, you bind netdata on localhost only (no internet access), and you ask the ssh client to tunnel port 19999 there to 19998 here, so that you can access netdata as `http://localhost:19998/`. This way you can just open an ssh session to all your servers and each netdata will appear on a different port on your computer.

2. run a firewall (you should have a firewall on all servers, even if they are not facing the internet - but especially to the ones facing the internet - firehol can be very helpful for this) and allow only your static IPs access netdata.

**Comment by jarmo** [comment]
@ktsao I totally agree that convenience is also important when using netdata, but can't agree that security is less important.

You're saying that netdata has read-only access, which is good as long as it doesn't have access to anything sensitive, but this is not always the case.

1) netdata configuration files are readable by everyone by default on Ubuntu - this means that when you have entered any passwords to any of the plugins configuration (postgres.conf for example) then possible attacker has access to everything in the database(s) - is there a preferred way how plugin configurations should be done so that at least non-netdata users won't be able to access configuration?

2) When netdata binds to external interfaces then it might come as a bad surprise to some users when heart-beat of a server is exposed. Google for netdata dashboards, there are more than one visible and I doubt that all administrators are aware of it happening;

3) When implementing SSL, authentication and authorization yourself, there's definitely are going to be security issues - let these things to be handled by software which are meant for it and keep netdata for monitoring only;

TLDR; My suggestions are to keep netdata only for monitoring and make it secure by default and print out some helpful messages about configuring it to be less-secure.

**Comment by ktsaou** [comment]
The issue you mention, about configuration files being world readable, is a problem. I'll fix that in the installer - thanks for mentioning it, but this is something distributions should handle using their policies.

Could you please pinpoint my comment where I implied that security is less important than convenience?

What I said, is that if a sysadmin has accidentally left netdata open on the internet, his security has a lot more issues than netdata. His infrastructure allows random ports on his servers being accessible on the internet, without any control and nothing we can do will solve this.

Let's consider this case. A sysadmin allows random ports on his server being accessible on the internet. Let's assume that netdata binds to `lo` by default.

So, this guy installs netdata and a message is telling him: "please edit /etc/netdata/netdata.conf and change the interface or IP netdata binds to, in order for it to be accessible remotely".

What you suggest, is that this guy, with the broken security, will magically think - when he sees the message above - to edit his firewall and block netdata. So, he has to be aware that his firewall allows random ports to be served on the internet and now, he understands that he needs to edit it and block just the netdata port.

I am afraid he will not. He is either not aware of the fact that his firewall has the wrong policy (firewalls should be: deny all, allow explicitly - and obviously his firewall is wrong), or he likes it the way it is. He will just edit netdata.conf, make the change and restart netdata.

The result is frustration for everyone and exactly the same security for him.

I agree that security is important (after all, I develop a firewall http://firehol.org, I maintain all the publicly available cybercrime feeds on the internet at http://iplists.firehol.org). But security has to be meaningful too and binding to `lo` by default, a server that is supposed to be accessed remotely, will have zero effect on security.

**Comment by jarmo** [comment]
Totally agree with your points there and since Ubuntu's firewall is off by default (as opposed to being on when assuming secure by default approach), then maybe since netdata might be the first service which is needed to be accessible from non-common ports (80, 443 and 22) might bring that firewall problem into attention.

Maybe there should be a security-related wiki page for netdata, which would cover all the sane things to do when using netdata (and it's plugins). For example nginx_log plugin does not have access to /var/log/nginx/access.log by default - how to solve it properly? One way would be to add netdata to adm group (under Ubuntu), but it seems to be too much.

What if that security page also had information about each plugin and how to configure them securely? For example, in addition to the nginx_log plugin problem, postgres (and MySQL as well) plugins both will have a plain text password in the configuration for "root" access. This in turn means that if netdata has a security problem, someone will get access to the whole database, which is definitely not good.

How should problems like these be solved from a security standpoint? Should there be a separate user for postgres and MySQL which would only have limited read access etc? Even I'm not sure right now how to solve these problems, but it seems that some plugins encourage to create possible security holes.

**Comment by ktsaou** [comment]
nice idea. I'll start something in this area and see how to goes. I'll post here when I have something.

**Comment by d4rkstar** [comment]
Hello,
i've found a quite simple and secure solution using nginx ([openresty](https://github.com/openresty/)) with lua support and [Nginx JWT](https://github.com/auth0/nginx-jwt)

The idea is this:

1. the user logins through a simple login page
2. the login page may setup an access token through jwt
3. after the login the user is authorized to see netdata (otherwise redirected to the login page)

One can create also other authentication methods without Nginx JWT using the [access_by_lua](https://github.com/openresty/lua-nginx-module#access_by_lua) configuration and custom scripts.

**Comment by tisc0** [comment]
I just workd it around with haproxy : very efficient.

userlist netdata_users
        user UserName insecure-password myInsecurePasswd

listen netdata
        mode tcp
        bind yourPublicIP:19999
        acl netdataAccess http_auth(netdata_users)
        http-request auth realm netdataStuff if !netdataAccess
        server local 127.0.0.1:19998

And uncomment and set the 'default port = 19998' in /opt/netdata/etc/netdata/netdata.conf 

but +1 for a little simple secure passwd included in netdata

Thanks a lot for that tool anyway, it's amazing.

**Comment by pjz** [comment]
I'm actually kind of -1 on building security into netdata.  It needs to focus on doing what it does the best that it can.  The problem with adding logins, https, etc etc to it is that it makes the code more complex, and more importantly, gives @ktsaou that much more code to maintain - when IMO he should just be focusing on netdata itself, instead of trying to keep up with the upgrade train as http/https evolve and change. 

If you really want this, though, I'd suggest a 'netdata-server' project that depends on and uses the existing netdata, but is independent enough that issues/development there doesn't end up blocking issues/development here.

**Comment by dcharbonnier** [comment]
the proxy feature on the new release 1.6 add more requirements for this feature, I don't think you can setup a "custom" authentication and activate the streaming

**Comment by ktsaou** [comment]
You are right.

You can bind netdata to multiple ports (add them to the `bind to` line). So, use one of them for proxying netdata through another web server and another one, firewall protected, for receiving metrics.

Example:

```
[web]
    bind to = 127.0.0.1:19999 *:19998
```

Using the above, netdata will listen at localhost port 19999 and on all IPs at port 19998. If you protect the second with a firewall to allow only the streaming hosts to send data, you are done.

**Comment by dcharbonnier** [comment]
I don't want/can't protect with a firewall, I would like to limit a port to the stream functionality.

**Comment by ktsaou** [comment]
Unfortunately, this the only option currently.

So, if netdata had the ability to accept streamed metrics on a dedicated port, would this be ok for you?

**Comment by dcharbonnier** [comment]
yes, that would allow to postpone the development of an authentication layer, with the new stream feature it's not possible anymore to proxy to add auth

**Comment by ktsaou** [comment]
ok, let's continue on #2149

**Comment by wbvalid** [comment]
Maybe it's also a good way to add rules in IPTABLES to restrict the IP(s) that are allowed to access the port 19999?

**Comment by MBlagui** [comment]
The problem if you have a shared hosting server with netdata even using nginx it won't protect it really! Why? Case one customer curl 127.0.0.1:19999 he will bypass nginx and will be able to access local resources. As long there is not even a token authentification it won't be secure.

Would be simple if we can a single token for authentification like a password so in this case any curl won't get a response.

**Comment by ktsaou** [comment]
@MBlagui the solution to this localhost problem is #2656 .

**Comment by gjpalau** [comment]
@ktsaou I've been trying to find the location of netdata.conf so I can add the 

```
[web]
	bind to = 10.1.1.1:19999 localhost:19999
```

and as such make it browsable only from my internal network.  But for the life of me I can't find the file... I have tried looking on `/opt/netdata/etc/netdata/netdata.conf` but all I get is this: # Get config from http://127.0.0.1:19999/netdata.conf

Could you provide me the step by step where to find this file so I can add that line?

Thanks!

**Comment by ktsaou** [comment]
The feature set of #2656 is in PR #2657 (not merge yet).

This will allow netdata and your web server to communicate via a unix domain socket (a `.sock` file on disk). You can protect this file (i.e. put it in a directory only your web server and netdata will have access) so that this communication will be completely private.

**Comment by MBlagui** [comment]
@ktsaou Thanks for the reply but still a token would be another solution. But I think it will require a lot more of work and not just requesting it.

I will check socks implementation once it's released.

**Comment by cakrit** [comment]
This issue is very old and mentions many things that have been available for a long time. For the title regarding authentication, monitor instead #6050.
